### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by 777777miSSU7777777**

### CODE REVIEW

File: app/api/auth.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package api
+
+import (
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/777777miSSU7777777/github-aggregator-web/app/util/http/cookie"
```
Comment (amaskalenka) [inline app/api/auth.go]: Please refer to [this](https://golang.org/doc/effective_go.html#names) paragraph and [that](https://blog.golang.org/package-names) article. Do not use names such `util`,  `common` for packages.

File: app/api/auth.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+//Auth Saves to cookie Github API access token from form on index page.
```
Comment (amaskalenka) [inline app/api/auth.go]: Please refer to [this](https://blog.golang.org/godoc-documenting-go-code) article as an example of how to provide meaningful and well formatted doc strings.

File: app/api/auth.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+func Auth(rw http.ResponseWriter, req *http.Request) {
+	if req.Method == http.MethodPost {
+		accessToken := req.FormValue("access_token")
+		log.Printf("Token: %s", accessToken)
```
Comment (amaskalenka) [inline app/api/auth.go]: You should not print sensitive data to log.

File: app/api/auth.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	"net/http"
+	"time"
+
+	"github.com/777777miSSU7777777/github-aggregator-web/app/util/http/cookie"
+)
+
+//Auth Saves to cookie Github API access token from form on index page.
+func Auth(rw http.ResponseWriter, req *http.Request) {
+	if req.Method == http.MethodPost {
+		accessToken := req.FormValue("access_token")
```
Comment (amaskalenka) [inline app/api/auth.go]: You must check on an empty string.

File: app/api/auth.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+		log.Printf("Token: %s", accessToken)
+		resp, err := http.Get("https://api.github.com/user?access_token=" + accessToken)
```
Comment (amaskalenka) [inline app/api/auth.go]: Please do not hardcode URL.

File: app/api/auth.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+		if err != nil {
```
Comment (amaskalenka) [inline app/api/auth.go]: You must return an HTTP 5xx error response with appropriate description.

File: app/api/auth.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+			log.Println(err.Error())
```
Comment (amaskalenka) [inline app/api/auth.go]: You do not need to call `Error` method explicitly.

File: app/api/auth.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+		}
+
+		if resp.StatusCode == 200 {
```
Comment (amaskalenka) [inline app/api/auth.go]: What if `http.Get` returns an error?

File: app/api/auth.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+			cookie.SaveCookie(rw, "access_token", accessToken, time.Hour)
```
Comment (amaskalenka) [inline app/api/auth.go]: You must not store sensitive data in plan format. Please encode token.

Comment (amaskalenka) [reply app/api/auth.go]: Why one hour as an expiration time for the cookie? Please do not hardcode an expiration timeout, make it configurable.

File: app/query/profile.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func getUserField(req *http.Request, key string) (interface{}, error) {
+	accessToken, err := cookie.GetCookieValue(req, "access_token")
+
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Get("https://api.github.com/user?access_token=" + accessToken)
```
Comment (amaskalenka) [inline app/query/profile.go]: Please do not hardcode URL

File: app/query/scopes.go [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+//GetScopes returns scopes for provided token.
+func GetScopes(req *http.Request) []string {
+	accessToken, err := cookie.GetCookieValue(req, "access_token")
+
+	if err != nil {
+		log.Println(err.Error())
+	}
+
+	resp, err := http.Get("https://api.github.com/user?access_token=" + accessToken)
```
Comment (amaskalenka) [inline app/query/scopes.go]: Do not you need to close `resp.Body`?

File: main.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+	"github.com/777777miSSU7777777/github-aggregator-web/app/api"
+	"github.com/777777miSSU7777777/github-aggregator-web/app/view/index"
+)
+
+var host string
+var port string
+
+func init() {
+	flag.StringVar(&host, "host", "127.0.0.1", "")
```
Comment (amaskalenka) [inline main.go]: Please add descriptions for CLI arguments.

File: main.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	flag.StringVar(&port, "p", "8080", "")
+	flag.Parse()
+}
+
+func main() {
+	log.Printf("Server started on %s:%s", host, port)
+	http.HandleFunc("/", index.Render)
+	http.HandleFunc("/auth", api.Auth)
+	http.HandleFunc("/logout", api.Logout)
+	err := http.ListenAndServe(fmt.Sprintf("%s:%s", host, port), nil)
```
Comment (amaskalenka) [inline main.go]:
```suggestion
	if err := http.ListenAndServe(fmt.Sprintf("%s:%s", host, port), nil); err != nil {
```

File: main.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+	if err != nil {
+		log.Println(err.Error())
```
Comment (amaskalenka) [inline main.go]: I would use `Fatal` here since it is really fatal error.

File: app/view/index/index.go [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+	accessToken, err := cookie.GetCookieValue(req, "access_token")
+
+	if err != nil {
+		log.Println(err.Error())
+	}
+
+	session.Authorized = accessToken != ""
+	if session.Authorized {
+		session.Username, _ = query.GetUsername(req)
```
Comment (amaskalenka) [inline app/view/index/index.go]: Why don't you check on errors?

File: app/util/http/header/header_util.go [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+func readHeader(hdr http.Header) map[string][]string {
+	header := map[string][]string{}
+
+	for key, value := range hdr {
+		header[key] = value
+	}
+
+	return hdr
```
Comment (amaskalenka) [inline app/util/http/header/header_util.go]: Why do you return what you receive?

File: app/query/profile.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+	"net/http"
+
+	"github.com/777777miSSU7777777/github-aggregator-web/app/util/http/body"
+	"github.com/777777miSSU7777777/github-aggregator-web/app/util/http/cookie"
+	"github.com/777777miSSU7777777/github-aggregator-web/app/util/json"
+)
+
+//GetUsername returns username from github for provided token.
+// Also returns error if happen.
+func GetUsername(req *http.Request) (string, error) {
```
Comment (amaskalenka) [inline app/query/profile.go]: Why are you issuing the same HTTP request three times to extract different parts of the same response?

File: app/query/profile.go [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+		return nil, err
+	}
+
+	encodedBody, err := json.BytesToMap(respBody)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return encodedBody[key], nil
```
Comment (zhiburt) [inline app/query/profile.go]: I think that's a bit semantic mistake
If key doesn't exist your code return `nil`
but is it `nil` means that this key doesn't exist in map or this is a correct value `nil`
may you better return some `error` like this `ErrNotFoundKey`

**Pull request merged**